CMakeLists.txt

```diff
@@ -1,179 +1,141 @@
-add_subdirectory(lib/gamepad)
-add_subdirectory(lib/gpio)
-add_subdirectory(lib/controllers)
-add_subdirectory(lib/neopixel)
-add_subdirectory(lib/pid)
-add_subdirectory(lib/sensors)
-add_subdirectory(lib/utility)
-
-if (NOT DEFINED APP)
-    message(FATAL_ERROR "APP variable not set in CACHE. Please invoke CMake with \"-DAPP=<app-name>\"")
-elseif (NOT EXISTS "${CMAKE_SOURCE_DIR}/apps/${APP}")
-    message(FATAL_ERROR "${APP} app does not exist in the apps folder")
-endif ()
-if (NOT DEFINED TARGET)
-    message(FATAL_ERROR "TARGET variable not set in CACHE. Please invoke CMake with \"-DTARGET=<target-name>\"")
-elseif (NOT EXISTS "${CMAKE_SOURCE_DIR}/targets/${TARGET}")
-    message(FATAL_ERROR "${TARGET} target does not exist in the targets folder")
-endif ()
-message(STATUS "Configuring and generating build for ${APP} app on ${TARGET} target")
-
-# Add CMakeList for target-specific files
-add_subdirectory(targets/${TARGET})
+# Project Name
+project(MarsRoverFirmware LANGUAGES C CXX ASM)
 
 # Add HW Bridge
 add_subdirectory(uwrt-mars-rover-hw-bridge)
 
-# Include Rover App CMakeLists
-add_subdirectory(apps/arm)
-add_subdirectory(apps/gamepad)
-add_subdirectory(apps/gimbal)
-add_subdirectory(apps/pdb)
-add_subdirectory(apps/science)
-
-# Include Test App CMakeLists
-add_subdirectory(apps/test-actuator-controller)
-add_subdirectory(apps/test-aeat-6012)
-add_subdirectory(apps/test-aeat-8800)
-add_subdirectory(apps/test-blinky)
-add_subdirectory(apps/test-blockingneo)
-add_subdirectory(apps/test-can)
-add_subdirectory(apps/test-mae3)
-add_subdirectory(apps/test-logger)
-add_subdirectory(apps/test-lookup-table)
-add_subdirectory(apps/test-adafruitSTEMMA)
-add_subdirectory(apps/test-pid)
-add_subdirectory(apps/test-pwm)
-add_subdirectory(apps/test-pwmin)
-add_subdirectory(apps/test-units)
\ No newline at end of file
+# Include libraries
+add_subdirectory(${LIBS_DIR}/actuators)
+add_subdirectory(${LIBS_DIR}/can)
+add_subdirectory(${LIBS_DIR}/encoder)
+add_subdirectory(${LIBS_DIR}/gamepad)
+add_subdirectory(${LIBS_DIR}/gpio)
+add_subdirectory(${LIBS_DIR}/controllers)
+add_subdirectory(${LIBS_DIR}/neopixel)
+add_subdirectory(${LIBS_DIR}/pid)
+add_subdirectory(${LIBS_DIR}/sensors)
+add_subdirectory(${LIBS_DIR}/utility)
+
+# Only add subdirectory of requested APP
+function(add_app_subdirectory APP_SUBDIRECTORY)
+    string(FIND ${APP_SUBDIRECTORY} / INDEX REVERSE)
+    math(EXPR INDEX "${INDEX} + 1")
+    string(SUBSTRING ${APP_SUBDIRECTORY} ${INDEX} -1 APP_NAME)
+
+    if ("${APP_NAME}" STREQUAL "")
+        message(FATAL_ERROR "add_app_subdirectory could not extract an app name from ${APP_SUBDIRECTORY}")
+    elseif ("$ENV{APP}" STREQUAL ${APP_NAME})
+        add_subdirectory(${APP_SUBDIRECTORY})
+    endif ()
+endfunction()
+
+## Include Rover Apps
+add_app_subdirectory(${APPS_DIR}/arm)
+add_app_subdirectory(${APPS_DIR}/gamepad)
+add_app_subdirectory(${APPS_DIR}/gimbal)
+add_app_subdirectory(${APPS_DIR}/pdb)
+add_app_subdirectory(${APPS_DIR}/science)
```

Comment:
What's the plan for if we might have different apps for different board revisions? Or are we just gonna feature flag things within the same app?

Reply:
If they close enough, I'd just keep a common app and feature flag it. else we can do a similar thing and revision the apps.

Reply:
but personally i think we want to move towards the single common app type of structure